### ISSUE TITLE
How can I set the base URL for an exchange?

### ISSUE THREAD

**Issue by quantitative-technologies** [ccxt/ccxt#19507]
### Operating System

Linux

### Programming Languages

Python

### CCXT Version

4.0.82

### Description

Apologies if this is the wrong place, but I could not find a "question" category.

My IP seems to be blocked by binance, specifically by `binance.com`. I contacted their support and they could not find any block, and suggested I use the alternate URL `https://www.binancezh.info/`.

I've read the Proxy section of the documentation but I don't think that is what I want.

Is there a setting for the base URL of an exchange, e.g. passed in as a configuration setting?

**Comment by carlosmiei** [comment]
Hello @quantitative-technologies, in this case I think this would be the easier way:

```Python

new_host = 'YOUR_VALUE_HERE'
exchange = ccxt.binance({
    'urls': {
        'api': {
            'wapi': f'https://{new_host}/wapi/v3',
            'sapi': f'https://{new_host}/sapi/v1',
            'sapiV2': f'https://{new_host}/sapi/v2',
            'sapiV3': f'https://{new_host}/sapi/v3',
            'sapiV4': f'https://{new_host}/sapi/v4',
            'dapiPublic': f'https://d{new_host}/dapi/v1',
            'dapiPrivate': f'https://d{new_host}/dapi/v1',
            'eapiPublic': f'https://e{new_host}/eapi/v1',
            'eapiPrivate': f'https://e{new_host}/eapi/v1',
            'dapiPrivateV2': f'https://d{new_host}/dapi/v2',
            'dapiData': f'https://d{new_host}/futures/data',
            'fapiPublic': f'https://f{new_host}/fapi/v1',
            'fapiPrivate': f'https://f{new_host}/fapi/v1',
            'fapiData': f'https://f{new_host}/futures/data',
            'fapiPrivateV2': f'https://f{new_host}/fapi/v2',
            'public': f'https://{new_host}/api/v3',
            'private': f'https://{new_host}/api/v3',
            'v1': f'https://{new_host}/api/v1',
            'papi': f'https://p{new_host}/papi/v1',
        },
    }
})
```

**Comment by quantitative-technologies** [comment]
I tried this. Not surprisingly, it works fine with the usual `new_host = api.binance.com` (my IP is apparently no longer blocked). 

However, when I tried the alternative `api.binancezh.info` I get an SSL certificate error:

```console
Exception has occurred: ExchangeNotAvailable       (note: full exception trace is shown but execution is paused at: _run_module_as_main)
binance GET https://api.binancezh.info/api/v3/exchangeInfo
  File "/home/user/.virtualenvs/binance-bot/lib/python3.9/site-packages/aiohttp/connector.py", line 980, in _wrap_create_connection
    return await self._loop.create_connection(*args, **kwargs)  # type: ignore[return-value]  # noqa
  File "/usr/lib/python3.9/asyncio/base_events.py", line 1081, in create_connection
    transport, protocol = await self._create_connection_transport(
  File "/usr/lib/python3.9/asyncio/base_events.py", line 1111, in _create_connection_transport
    await waiter
  File "/usr/lib/python3.9/asyncio/sslproto.py", line 528, in data_received
    ssldata, appdata = self._sslpipe.feed_ssldata(data)
  File "/usr/lib/python3.9/asyncio/sslproto.py", line 188, in feed_ssldata
    self._sslobj.do_handshake()
  File "/usr/lib/python3.9/ssl.py", line 944, in do_handshake
    self._sslobj.do_handshake()
ssl.SSLCertVerificationError: [SSL: CERTIFICATE_VERIFY_FAILED] certificate verify failed: Hostname mismatch, certificate is not valid for 'api.binancezh.info'. (_ssl.c:1123)
```

**Comment by quantitative-technologies** [comment]
Can I please get left with the SSL certificate issue? Should I open a new issue?

I was using a VPN with the "built-in" URL, but now binance is blocking that too.

**Comment by carlosmiei** [comment]
@quantitative-technologies I'm not sure that issue is related to ccxt directly.

**Comment by carlotestor** [comment]
Yes — pass a `urls` override at construction, exactly as @carlosmiei showed. For Binance specifically there's also a shorter way, since all its hosts are derived from a single `hostname` property:

```python
exchange = ccxt.binance({
    'hostname': 'binancezh.info',    # replaces binance.com in api./fapi./dapi./eapi. URLs
})
```
That rewrites `api.binancezh.info`, `fapi.binancezh.info`, etc. in one shot, instead of listing every sub-API by hand.

On the follow-up SSL error: that's not a CCXT issue — those mirror domains frequently serve a certificate that doesn't match the hostname (they're intended for use with Binance's own clients). Your options are:
- Verify with `openssl s_client -connect api.binancezh.info:443 -servername api.binancezh.info` to see what CN/SAN it presents. If the cert genuinely doesn't cover the host, no client will accept it.
- Use a proxy/VPN with the official `api.binance.com` instead (`exchange.https_proxy = '...'`), which is the supported path and keeps TLS valid.

Disabling certificate verification is possible but strongly discouraged for an authenticated trading connection — MITM on that channel means stolen keys.
